crates/sui-core/src/authority.rs

```diff
@@ -3738,6 +3740,96 @@ impl AuthorityState {
         (next_protocol_version, system_packages)
     }
 
+    /// Creates and executes an authenticator state tx, but does not commit the
+    /// results to the db.
+    ///
+    /// If the authenticator state object does not yet exist, it creates an
+    /// AuthenticatorStateCreate tx. Otherwise, it creates an AuthenticatorStateExpire tx.
+    ///
+    /// The effects of the tx are only written to the database after a certified
+    /// checkpoint has been formed and executed by CheckpointExecutor
+    pub async fn create_and_execute_authenticator_state_tx(
+        &self,
+        epoch_store: &Arc<AuthorityPerEpochStore>,
+        checkpoint: CheckpointSequenceNumber,
+    ) -> anyhow::Result<Option<TransactionEffects>> {
+        if !epoch_store.protocol_config().enable_jwk_consensus_updates() {
+            return Ok(None);
+        }
+
+        let authenticator_state_exists = epoch_store.authenticator_state_exists();
+        let tx = if authenticator_state_exists {
```

Comment:
I think this is actually problematic given how authenticator_state_exists is cached in the epoch store. Since it's cached in the constructor, instead of truly being the state at the beginning of epoch, it not crash-safe.
You can have the object created and committed (when checkpoint executor executed this creation tx), then the node crashes, and when it wakes up it would think the object exists, then the validator path started to process the last checkpoint and it would fork.
The proper fix is likely to cache it in epoch start config.

Reply:
I'm confused - this was exactly the case I'm trying to prevent here. We don't re-check the object existence, we read the cached value from the db.

Reply:
I read this a bit more, and I think this should work - we snapshot whether object exists on the epoch store creation(so the function name is in fact confusing, it checks not that object exists, it checks that it existed in the beginning of the epoch), so it is essentially same as having a field in the EpochStartConfiguration.

I agree with the other comment from @lxfind that it probably would be cleaner to just put the field there, instead of creating new table though.

Reply:
Yes, I got confused and thought that EpochStartConfig was not versioned, and so using that would be difficult. But it is versioned, so i'm making that change right now.

Reply:
done